### PR TITLE
Introduction of remote gradnorm downsampling

### PR DESCRIPTION
**Pull request by francescodeaglio**
Data selection policy based on gradient norm (based on idiap/EPFL importance sampling)

### CODE REVIEW

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsample_strategy import (
+    AbstractRemoteDownsamplingStrategy,
+)
+
+
+class RemoteGradNormDownsampling(AbstractRemoteDownsamplingStrategy):
+    """
+    Method adapted from
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py:11]: Is it adapted or identical? If adapted, can you add a sentence explaining what is different from the paper?

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+        self.per_sample_loss_fct = per_sample_loss
+
+    def get_probabilities(self, forward_output: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        if isinstance(self.per_sample_loss_fct, torch.nn.CrossEntropyLoss):
+            # no need to autograd if cross entropy loss is used since closed form solution exists.
+            with torch.inference_mode():
+                # Because CrossEntropyLoss includes the softmax, we need to apply the
+                # softmax to the forward output to obtain the probabilities
+                probs = torch.nn.functional.softmax(forward_output, dim=1)
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py:29]: Could you add a brief comment while dim = 1?

**Review by MaxiBoether — APPROVED** [review]
Thank you! I have had one small comment - after addressing it and merging main (or not, not sure whether it is applicable) feel free to merge this! Thank you!

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py (hunk 1 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py:11]: Other optimisations are also proposed in the paper, but do not make sense in the context of dynamics datasets. This reproduces the basic version.

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py (hunk 2 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py:29]: You want to softmax the logits, not the batch

File: modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py (hunk 1 above)
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/remote_gradnorm_downsample.py:11]: ok I meant whether you changed the basic version from the paper or not. but if i understand you correctly, you didn't :)

**Comment by MaxiBoether** [comment]
feel free to merge
